### PR TITLE
DEV-1389 Add optional free space check

### PR DESCRIPTION
**Pull request by spacid**
Add an optional check, that checks if there is enough free space on the
target server before transferring the file.

Also, refactor the paramiko client as a instance variable, so that the
client can be reused. This also reduces code copy.

### CODE REVIEW

**Comment by spacid** [comment]
There is a warm shutdown implemented, in the sense that if you send the TERM signal, the service will wait until the transfer is finished. However, the free space check is blocking, meaning that it will wait indefinitely until the space is freed. In the worst case scenario, this means that the service will never stop. In the deployment environment however, there is timeout parameter. If the service hasn't stopped until then, it will terminate the service.

TODO: Adapt the warm shutdown, that when the service is in a free-space-check loop, it just shut downs, sending a NACK (requeue=True) to RABBITMQ and closing the connection form the client side.

**Review by RudolfDG — APPROVED** [review]

File: docs/s3-t-s_sequence-diagram.plantuml [hunk 1]
```diff
@@ -14,11 +14,18 @@ sts -> rabbit: Listen to queue
 loop
     rabbit -> sts: Transfer message request
     sts -> tra: Connect via SSH
+    opt
+        tra -> tra: Check free space
+        loop While not enough free space
+                tra -> tra: Sleep 120 seconds
+                tra -> tra: Check free space
```
Comment (maartends) [inline docs/s3-t-s_sequence-diagram.plantuml]:
```suggestion
                tra -> tra: Check free space
                tra -> tra: Sleep 120 seconds
```

File: README.md [hunk 2]
```diff
@@ -9,6 +9,8 @@ We don't want any traffic in the service itself, so the file is transferred from
 A transfer will be split up in multiple parts of the same size. Each part will be transferred simultaneously, running in a separate thread. Afterwards
 the file will be assembled into the destination file.
 
+There is an optional free space check in which the remote server needs to have a certain amount of free space (in percentage) before the file is transferred. If that free space is not met, the service will sleep indefinitely until the space is freed. The information needed for the check is defined by the ENV vars: `SSH_FREE_SPACE_PERCENTAGE`, `SSH_FILE_SYSTEM`. The vars are mandatory but the values may be empty. In fact, if at least one of the values are empty than the check will not be executed, and the file will be transferred regardless of the free space on the target server.
```
Comment (maartends) [inline README.md]: The value needed for `SSH_FILE_SYSTEM` might need some more explanation (I assume it has to be the mountpoint?).

**Review by maartends — APPROVED** [review]
For now, `_check_free_space` will hang indefinitely until space is freed on the remote server. This also without warning (except the log.info "Free space: [...] Space needed: [...]").
Could be a future improvement.

+1 for README and diagram-updates.

File: docs/s3-t-s_sequence-diagram.plantuml (hunk 1 above)
Comment (spacid) [reply docs/s3-t-s_sequence-diagram.plantuml]: In the diagram there is a check first if there is enough free space before this loop:
image

But perhaps this is not the most clear way to visualize the functionality?

**Review by maartends — APPROVED** [review]

**Pull request merged**